src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs

```diff
@@ -372,34 +372,42 @@ private BoundStatement UpdateStatement(BoundSpillSequenceBuilder builder, BoundS
 
                     case BoundKind.FieldAccess:
                         var field = (BoundFieldAccess)expression;
-                        if (field.FieldSymbol.IsReadOnly)
+                        var fiedlSymbol = field.FieldSymbol;
+                        if (fiedlSymbol.IsStatic)
                         {
-                            if (field.FieldSymbol.IsStatic) return field;
-                            if (field.FieldSymbol.ContainingType.IsValueType) goto default;
-                            // save the receiver; can get the field later.
-                            var receiver = Spill(builder, field.ReceiverOpt, (refKind != RefKind.None && field.FieldSymbol.Type.IsReferenceType) ? refKind : RefKind.None, sideEffectsOnly);
-                            return field.Update(receiver, field.FieldSymbol, field.ConstantValueOpt, field.ResultKind, field.Type);
+                            // no need to spill static fields if used as locations or if readonly
+                            if (refKind != RefKind.None || fiedlSymbol.IsReadOnly)
+                            {
+                                return field;
+                            }
+                            goto default;
                         }
-                        goto default;
+
+                        if (refKind == RefKind.None) goto default;
+
+                        var receiver = Spill(builder, field.ReceiverOpt, field.FieldSymbol.ContainingType.IsValueType ? refKind : RefKind.None);
+                        return field.Update(receiver, field.FieldSymbol, field.ConstantValueOpt, field.ResultKind, field.Type);
 
                     case BoundKind.Call:
                         var call = (BoundCall)expression;
-                        if (refKind != RefKind.None)
+                        if (refKind != RefKind.None && refKind != RefKind.RefReadOnly)
                         {
                             Debug.Assert(call.Method.RefKind != RefKind.None);
                             _F.Diagnostics.Add(ErrorCode.ERR_RefReturningCallAndAwait, _F.Syntax.Location, call.Method);
-                            refKind = RefKind.None; // Switch the RefKind to avoid asserting later in the pipeline
                         }
+                        // method call is not referentially transparent, we can only spill the result value. 
+                        refKind = RefKind.None; 
                         goto default;
 
                     case BoundKind.ConditionalOperator:
                         var conditional = (BoundConditionalOperator)expression;
-                        if (refKind != RefKind.None)
+                        if (refKind != RefKind.None && refKind != RefKind.RefReadOnly)
```

Comment:
Logged [https://github.com/dotnet/roslyn/issues/20890](https://github.com/dotnet/roslyn/issues/20890) to check on the test coverage. Also curious what VB does. While VB does not produce ref-returning methods it can consume them, so this would be applicable to VB as well.

---
In reply to: [127526424](https://github.com/dotnet/roslyn/pull/20883#discussion_r127526424) [](ancestors = 127526424,127526157,127518317)